Request from Workaholics team

ivandkoz opened the pull request:
The Workaholics presents new protein analysis tool.
The Workaholics team members:
Ivan Kozin
Dasha Sokolova 
Yulia Volkova

ivandkoz committed:
Create protein_analysis_tool.py

Added new empty file protein_analysis_tool.py

stegodasha committed:
Add function get_amino_acid_sum

stegodasha committed:
Add function codon_optimization

ivandkoz committed:
Merge pull request #1 from stegodasha/Sokolova_Dasha

Sokolova dasha

yvolko committed:
Add main protein function, validate, molecular weight and one letter to three code functions

yvolko committed:
Add readme draft

yvolko committed:
Correct input and output data types of functions

yvolko committed:
Fixed new line issue

stegodasha committed:
Add input of cell type for user

stegodasha committed:
Fixed a bug in the function codon_optimization and add a mouse for optimization

stegodasha committed:
Add docstrings and typing to functions

stegodasha committed:
Added an English translation

stegodasha committed:
Remove all unnecessary

stegodasha committed:
Merge pull request #2 from stegodasha/Sokolova_Dasha

Sokolova dasha

ivandkoz committed:
Added functions: lenght,  reverse, name_transform,
check_amino_acid

yvolko committed:
Add functions get_amino_acid_sum and codon_optimization to main

ivandkoz committed:
Merge branch 'main' into HW_4_Dasha

ivandkoz committed:
Merge pull request #3 from ivandkoz/HW_4_Dasha

Hw 4 dasha merged to main

ivandkoz committed:
Merge branch 'main' into HW_4_Yulia

ivandkoz committed:
Merge pull request #4 from ivandkoz/HW_4_Yulia

Hw 4 yulia merged to main

ivandkoz committed:
Update protein_analysis_tool.py

Debug amino_names_transform. Debug protein_analysis input

ivandkoz committed:
Update protein_analysis_tool.py

Redisign functions molecular_weight and one_letter_to_three. Dictionaries removed from functions

ivandkoz committed:
Update protein_analysis_tool.py

Add return to get_amino_acid_sum and correct output format

ivandkoz committed:
Update protein_analysis_tool.py

Delete functions beautiful_print and reverse. Redesign codon_optimization output

ivandkoz committed:
Update protein_analysis_tool.py

Add support lenght function

ivandkoz committed:
Update protein_analysis_tool.py

Delete print from get_amino_acid_sum function

ivandkoz committed:
Update protein_analysis_tool.py

Rename 'lenght' to 'length'

stegodasha committed:
Update README.md

Add to list of procedures get_amino_acid_sum and codon_optimization

stegodasha committed:
Fixed docstrings for functions codon_optimization and get_amino_acid_sum in protein_analysis_tool.py

ivandkoz committed:
Update README.md

Add contact information

ivandkoz committed:
Update README.md

Add team photo

stegodasha committed:
Fixed docstrings for functions codon_optimization and get _amino_acid_sum and  delete the excess in protein_analysis_tool.py

ivandkoz committed:
Update README.md

Debug new lines problems in contacts

ivandkoz committed:
Update README.md

Debug newlines

ivandkoz committed:
Update README.md

Newlines debug

yvolko committed:
Update protein_analysis_tool.py

Add arguments and returns to docstring of main, correct arguments types

ivandkoz committed:
Update protein_analysis_tool.py

Add input types to length and check_amino_acid functions

ivandkoz committed:
Update protein_analysis_tool.py

Correct input codon_optimization in main function

ivandkoz committed:
Update protein_analysis_tool.py

Redesign codon optimization error message

stegodasha committed:
Add example of use function get_amino_acid_sum and codon_optimization

ivandkoz committed:
Update protein_analysis_tool.py

Add amino_brutto dictionary and brutto_count function

ivandkoz committed:
Update protein_analysis_tool.py

Debug input types

yvolko committed:
Update README.md

Make corrections to adjust readme to actual code

yvolko committed:
Update README.md

Make examples of use prettier

ivandkoz committed:
Update protein_analysis_tool.py

Restructured main function (protein_analysis). Add new way to call other functions

ivandkoz committed:
Update protein_analysis_tool.py

Change errors message in check_amino_acid and name_transform functions

ivandkoz committed:
Update protein_analysis_tool.py

moved dictionaries from functions to the beginning of the code

ivandkoz committed:
Update README.md

Change design to python code in "examples of use" and "Possible errors"

ivandkoz committed:
Update README.md

add ")" in python code

ivandkoz committed:
Update protein_analysis_tool.py

Rename "lenght" to "length"

ivandkoz committed:
Update README.md

Newlines debug

ivandkoz committed:
Update README.md

Newlines debug

yvolko committed:
Update protein_analysis_tool.py

Fix length name in procedure dictionary

ivandkoz committed:
Update README.md

Debug comments in python code

ivandkoz committed:
Update README.md

Comments debug

yvolko committed:
Update README.md

Add personal contribution

yvolko committed:
Update README.md

Add new lines to presonal contribution

yvolko committed:
Update README.md

make contribution part pretty

yvolko committed:
Update README.md

correct misspelling

ivandkoz committed:
Update protein_analysis_tool.py

Made default format 1. Add check for possible incorrect format specification. Add functions check_amino_acid_three_letter, check_length

ivandkoz committed:
Update README.md

Add two func into Ivan func list

ivandkoz committed:
Update README.md

Delete unnecessary "```"

yvolko committed:
Update protein_analysis_tool.py

Change arguments and outputs types

ivandkoz committed:
Update README.md

Update "how to use"

ivandkoz committed:
Update README.md

Newlines debug

yvolko committed:
Update protein_analysis_tool.py

Change argument in name_transform

yvolko committed:
Update protein_analysis_tool.py

Make strings in docstring shorter

yvolko committed:
Update README.md

Add info about brutto count to general information

ivandkoz committed:
Update protein_analysis_tool.py

Delete unnecessary returns from name_transform function

ivandkoz committed:
Update README.md

Newlines debug

yvolko committed:
Update protein_analysis_tool.py

Add docstring to some functions

yvolko committed:
Update protein_analysis_tool.py

Fix indent issue

ivandkoz committed:
Update protein_analysis_tool.py

Tabs debug

yvolko committed:
Update protein_analysis_tool.py

Change name replacer_Mouse to replacer_mouse

yvolko committed:
Update protein_analysis_tool.py

Renamed format to letter_format

ivandkoz committed:
Update README.md

Update "how to use"

yvolko committed:
Update README.md

Add examples of use

yvolko committed:
Update README.md

Add more erros

ivandkoz committed:
Update protein_analysis_tool.py

Code design

ivandkoz committed:
Update protein_analysis_tool.py

Return debug

yvolko committed:
Update protein_analysis_tool.py

Renamed bool functions

ivandkoz committed:
Update README.md

Debug erros output

yvolko committed:
Update README.md

Changed functions names in contribution part

ivandkoz committed:
Update protein_analysis_tool.py

Redesign error message in codon_optimization

yvolko committed:
Update protein_analysis_tool.py

Change error message of codon optimisation

yvolko committed:
Update protein_analysis_tool.py

Change Note to Warning

ivandkoz committed:
Update README.md

Add new chapter

ivandkoz committed:
Update README.md

Add new errors message description

yvolko committed:
Update README.md

Correct general info

ivandkoz committed:
Update README.md

last update

ivandkoz committed:
Update README.md

Latest update

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    - List of of floats corresponding to the molecular weight in kDa
+    """
+    molecular_weights = []
+    for seq in amino_acid_seqs:
```
Цикл for лучше написать в главной функции. Сейчас приходится в каждой функции его писать, это ненужный повтор кода.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+            "V": 0,
+            "W": 0,
+            "Y": 0,
+        }
```
Лучше было погуглить и создать словарь с 0 на лету. Пример:
```suggestion
        amino_acid_count = dict([(key, 0) for key in amino_short_names_dic.keys()])
```

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+            codon_optimization_ecoli += [
+                "".join([replacer_ecoli(n, n) for n in protein_sequences[amino_acid]])
+            ]
+        return codon_optimization_ecoli
```
К этой функции есть ряд замечаний:
- Допустимые клеточные линии лучше было записать в отдельную переменную, потому что если видов 1000, а пользователь вводит не представленный вид, то сперва код пройдет 1000 if/else.
- replacer_ecoli = ecoli_triplets.get. Не нужный этап. Вы можете напрямую обратиться к требуемому словарю.
- for amino_acid in range(len(protein_sequences)). Код внутри for повторяется трижды, лучше было тогда вынести логику в for в отдельную функцию. Сейчас идет ненужное повторение кода.
- Замечание по тому же циклу for. Непонятно, зачем нужно иттерироваться по индексу аминокислотного остатка, можно было просто иттерироваться по аминокислотам.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    Return:
+    - list of int values corresponding to the length of sequences"""
+    result = [len(seq) for seq in seqs]
+    return result
```
Работает верно, но можно было выдумать более интересную функцию.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    test_three_letters = []
+    if letter_format == 1:
+        for seq in seqs:
+            multiple_of_three.append(is_length_divisible_by_3(seq))
+            test_three_letters.append(is_amino_acid_three_letter(seq))
+            seq = seq.upper()
+            for letter in seq:
```
Недоработана логика. Если letter_format = 1, а пользователь подает аминокислотную последовательность в 3х буквенной записи, то вернется True. Но потом это знание не используется в нижестоящем for и 3х буквенный аминокислотный остаток воспринимается как 3 аминокислоты.  Если вы хотели переводить 3-буквенную запись в однобуквенную, то нужно было в for сделать шаг = 3. Либо же нужно было выбрасывать исключение, что пользователь подает в 3-буквенной записи последовательность, а letter_format == 1.

ivandkoz replied:
Спасибо за замечание. Это бутылочное горлышко нашего проекта. Проблема в том, что иногда нельзя выяснить какую последовательность вводит пользователь. Например, если пользователь хочет ввести трехбуквеную последовательность LysLysHis, но при этом указывает letter_format = 1, то встает вопрос -- как понять, это ошибка в указании letter_format и ввод трехбуквенный или ввод все же однобуквенный (в данной последовательности каждая буква является элементом однобуквенного кода)? 

Мы придумали решение, которое не решает однозначно проблему, но тем не менее может помочь избежать ошибки интерпретации при указании letter_format = 1. Мы проверяем каждую последовательность на кратность трем ее длины и на то, являются ли каждые "триплеты" последовательности трехбуквенным кодом аминокислот. При совпадении всех условий во всех заданных последовательностях, программа выводит сообщение с предупреждением : "Warning: all your sequences are similar to three-letter ones. Check the letter_format value"

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    if len(input_amino) == 1:
+        letter = input_amino
+        if letter not in amino_short_names_dic.keys():
+            raise ValueError(f"Error {letter} is not an amino acid. Correct your input")
```
Не самое удачное решение. Функция начинается c `is_`, исходя из названия ожидается, что функция вернет True/False. А она возвращает True или ошибку ValueError. Лучше было бы возвращать True/False,  а в функции, из которой вызывается is_amino_acid уже сделать проверку, что если функция вернула False, вызвать ошибку.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+        brutto = [sum(i) for i in brutto_pair]
+        brutto_dict = dict(zip(elements, brutto))
+        result.append(brutto_dict)
+    return result
```
Хорошая функция.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    if seq_len % 3 == 0:
+        return True
+    else:
+        return False
```
Можно убрать else, потому что если код не зайдет в if, то вернет False. А если зайдет в if, вернется True и не будет уже исполнять следующую строку в коде.
```suggestion
    return False
```

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+    elif procedure == "codon_optimization":
+        return procedures.get(procedure)(amino_acid_seqs, cell_type)
+    else:
+        return procedures.get(procedure)(amino_acid_seqs)
```
👍

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+def protein_analysis(
+    *args: str, procedure: str, cell_type: str = None, letter_format: int = 1
+) -> list:
+    """
+    Function protein_analysis:
+    - calculates predicted molecular weight of amino acid sequences in kDa (procedure name: molecular_weight)
+    - translate aa sequences from one-letter to three-letter code (procedure name: one_letter_to_three)
+    - calculates total amount of each amino acid in the sequences (procedure name: get_amino_acid_sum)
+    - makes DNA based codon optimization for the introduced amino acid sequences, support 3 types of cells:
+      Esherichia coli, Pichia pastoris, Mouse (procedure name: codon_optimization)
+    - calculates length of amino acid sequences (procedure name: length)
+    - counts the number of atoms of each type in a sequence (procedure name: brutto_count)
+
+    Arguments:
+    - one or multiple string of protein sequences written one letter or three letter code (not mixed)
+    - name of procedure as string
+    - cell type (required only for codon_optimization procedure)
+    - letter_format of code for the protein sequences as int: 1 for one letter, 3 for three letter code
+
+    Return:
+    - molecular_weight procedure returns list of floats
+    - one_letter_to_three procedure returns list of strings
+    - get_amino_acid_sum procedure returns list of dictionaries
+    - codon_optimization procedure returns list of strings
+    - length procedure returns list of int values
+    - brutto_count procedure returns list of dictionaries with counts of atoms in the sequence
+    """
```
Докстринг описано подробно, но носит более произвольных характер. Посмотрите в документации Python, как правильно его оформлять. Кстати, у вас есть хороший пример в функции molecular_weight

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+}
+
+
+def protein_analysis(
```
Обычно главная функция пишется ниже тех, которые из нее вызываются.

albidgy commented on protein_analysis_tool.py:
```diff
@@ -0,0 +1,491 @@
+amino_short_names_dic = {
```
Имена констант пишутся заглавными буквами.

albidgy reviewed (commented):
Хорошая работа!
__Плюсы:__
- README написан подробно. Есть только 1 замечание: нигде не указано, как подавать аминокислотную последовательность в 3-буквенном формате.
- Код в целом написан аккуратно, видно, что вы поработали над стилем кода.
- Хороший нейминг.
- Здорово, что все доступные функции записали в словарь, смотрится аккуратно.
- Замечаний по комментариям к коммитам у меня нет, понятно, что было добавлено в каждом коммите. Я только могу посоветовать писать их в одном стиле, когда работаете в команде.

__Замечания:__
- Не хватает единообразия в названии переменных. Где-то переменная с аминокислотами называется protein_sequences, где-то seqs. Лучше выбрать одно название и использовать его во всех функциях.
- Старайтесь избегать написания одних и тех же блоков кода.

__Баллы:__
- README - 2.5 балла (+0.2 за фото, -0.2 за отсутствие описания, как подавать аминокислоты в трехбуквенном формате).
- Функции - 1.5 * 5 = 7.5 баллов
- штрафы: функция name_transform не доделана -0.7, ~~повторение блоков кода -1~~, константы не с заглавной буквы -0.1, функция is_amino_acid возвращает не True/False -0.2 = -2 балла

__Итого:__ 9 баллов.

stegodasha committed:
Update amino_acid_sum and codon_optimization

ivandkoz committed:
Update protein_analysis_tool.py

The outputs of the is_amino_acid function have been changed to True and False. Removed unimportant "elses"

ivandkoz committed:
Update protein_analysis_tool.py

Changing case in constants to uppercase

albidgy commented:
Я пока пересматривала вашу работу заметила, что сильно штрафанула вас за повторение блоков кода. Хочу поднять балл до 9.
